File: go/luhn/luhn.go
package luhn

import (
	"strconv"
)

func Valid(input string) bool {

	if len(input) <= 1 {
		return false
	}

	var sum int
	var chars int
	for _, c := range reverse(input) {
		if c == ' ' {
			continue
		}

		chars++
		v, err := strconv.Atoi(string(c))
		if err != nil {
			return false
		}

		if chars%2 == 0 {
			sum += double(v)
			continue
		}
		sum += v
	}

	if chars <= 1 {
		return false
	}

	return sum%10 == 0
}

func double(in int) int {
	double := in * 2
	if double < 9 {
		return double
	}
	return double - 9
}

func reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

File: ruby/luhn/luhn.rb
module LuhnableString
  refine String do
    private

    def valid_length?(id_number)
      id_number.length > 1
    end

    def digits_only?(id_number)
      !id_number.match(/\D/)
    end

    def valid_sum?(id_number)
      (sum(id_number) % 10).zero?
    end

    def sum(id_number)
      digits = id_number.chars.map(&:to_i).reverse
      digits.map.with_index do |digit, index|
        index.odd? ? multiply(digit) : digit
      end.sum
    end

    def multiply(digit)
      value = digit * 2
      value > 9 ? value - 9 : value
    end

    public

    def valid_luhn?
      id_number = gsub(/\s+/, '')
      valid_length?(id_number) && digits_only?(id_number) && valid_sum?(id_number)
    end
  end
end

module Luhn
  using LuhnableString

  def self.valid?(id_number)
    id_number.valid_luhn?
  end
end
